File: csharp/14_abstraction.cs
// ============================================================
// Topic:  Abstraction
// Summary: An abstract base defines a contract; it can't be instantiated.
// Run:     dotnet script csharp/14_abstraction.cs
// ============================================================

using System;
using System.Globalization;

Console.WriteLine(new Circle(2).Describe()); // Output: Circle with radius 2 has area 12.57
Console.WriteLine(new Square(3).Describe()); // Output: Square with side 3 has area 9.00
// new Shape(); // error: Cannot create an instance of the abstract type 'Shape'

// ----- 1. Abstract class: declares WHAT, not HOW -----
abstract class Shape
{
    public abstract double Area();   // no body = subclasses must implement
    public abstract string Label();

    // concrete method built on the abstract ones
    public string Describe() =>
        $"{Label()} has area {Area().ToString("F2", CultureInfo.InvariantCulture)}";
}

// ----- 2. Concrete subclasses MUST implement every abstract member -----
class Circle : Shape
{
    private int r;
    public Circle(int r) { this.r = r; }
    public override double Area() => Math.PI * r * r;
    public override string Label() => $"Circle with radius {r}";
}

class Square : Shape
{
    private int s;
    public Square(int s) { this.s = s; }
    public override double Area() => s * s;
    public override string Label() => $"Square with side {s}";
}

// Cross-language note:
//   C# has `abstract class` (shown — may mix abstract + concrete members and state)
//   and `interface` (pure contract). `abstract`/`override` are compiler-enforced: a
//   concrete subclass that skips an abstract member won't compile.

File: java/14_abstraction.java
// ============================================================
// Topic:  Abstraction
// Summary: An abstract base defines a contract; it can't be instantiated.
// Run:     java java/14_abstraction.java   (Java 11+ single-file launch)
// ============================================================

import java.util.Locale;

public class Abstraction {
    public static void main(String[] args) {
        System.out.println(new Circle(2).describe()); // Circle with radius 2 has area 12.57
        System.out.println(new Square(3).describe()); // Square with side 3 has area 9.00
        // new Shape(); // error: Shape is abstract; cannot be instantiated
    }
}

// ----- 1. Abstract class: declares WHAT, not HOW -----
abstract class Shape {
    abstract double area();   // no body = subclasses must implement
    abstract String label();

    // concrete method built on the abstract ones
    String describe() {
        return String.format(Locale.US, "%s has area %.2f", label(), area());
    }
}

// ----- 2. Concrete subclasses MUST implement every abstract method -----
class Circle extends Shape {
    private int r;
    Circle(int r) { this.r = r; }
    @Override double area() { return Math.PI * r * r; }
    @Override String label() { return "Circle with radius " + r; }
}

class Square extends Shape {
    private int s;
    Square(int s) { this.s = s; }
    @Override double area() { return s * s; }
    @Override String label() { return "Square with side " + s; }
}

// Cross-language note:
//   Java has `abstract class` (shown — mixes abstract + concrete members and state)
//   and `interface` (a pure contract; since Java 8 it may also carry `default` methods).
//   A concrete subclass that skips an abstract method won't compile.
